### PR TITLE
Add Match Status information to the Field Monitor

### PR DESCRIPTION
**Pull request by FletcherS7**
Added Current Match, Match Time, Match State to Field Monitor. Added Red and Blue score to FTA Field Monitor.

Created groundwork for adding robot bandwidth.

These changes were tested at Capital City Classic.

### CODE REVIEW

File: field/driver_station_connection.go [hunk 1]
```diff
@@ -41,6 +41,7 @@ type DriverStationConnection struct {
 	BatteryVoltage            float64
 	DsRobotTripTimeMs         int
 	MissedPacketCount         int
+	Bandwidth                 float32
```
Comment (patfair) [inline field/driver_station_connection.go:44]: What unit is this in? Can you include it in the name? (e.g. `BandwidthMbps`)

File: field/driver_station_connection.go [hunk 2]
```diff
@@ -107,6 +108,17 @@ func (arena *Arena) listenForDsUdpPackets() {
...
+			        rawBw := data[9:11]
+
+			        bw := uint16(rawBw[0]) << 8
+			        bw |= uint16(rawBw[1])
+
+			        dsConn.Bandwidth = float32(bw) / 256.0
+
+
+
+
```
Comment (patfair) [inline field/driver_station_connection.go:121]: Remove trailing empty lines

File: field/driver_station_connection.go [hunk 3]
```diff
@@ -271,6 +283,7 @@ func (dsConn *DriverStationConnection) decodeStatusPacket(data [36]byte) {
 
 	// Number of missed packets sent from the DS to the robot.
 	dsConn.MissedPacketCount = int(data[2]) - dsConn.missedPacketOffset
+
```
Comment (patfair) [inline field/driver_station_connection.go:286]: Remove trailing empty line

File: static/js/field_monitor_display.js [hunk 4]
```diff
@@ -4,12 +4,20 @@
...
 
-// Handles a websocket message to update the team connection status.
+
 var handleArenaStatus = function(data) {
+  // If getting data for the wrong match (e.g. after a server restart), reload the page.
+  if (currentMatchId == null) {
+    currentMatchId = data.MatchId;
+  } else if (currentMatchId !== data.MatchId) {
+    location.reload();
+  }
```
Comment (patfair) [inline static/js/field_monitor_display.js:19]: Is this not happening automatically? In theory, when the page loses its websocket connection to the server it should retry connecting every three seconds, and when it reconnects it should get all the events that populate the initial state.

File: static/js/field_monitor_display.js [hunk 5]
```diff
@@ -86,11 +95,20 @@ var handleArenaStatus = function(data) {
       } else {
         teamRobotElement.text(dsConn.BatteryVoltage.toFixed(1) + "V");
       }
+      console.log(dsConn.Bandwidth);
```
Comment (patfair) [inline static/js/field_monitor_display.js:98]: Remove log if it was just used for debugging

File: static/js/field_monitor_display.js [hunk 6]
```diff
@@ -27,6 +35,7 @@ var handleArenaStatus = function(data) {
     var teamRadioTextElement = $(teamElementPrefix + "Radio span");
     var teamRobotElement = $(teamElementPrefix + "Robot");
     var teamBypassElement = $(teamElementPrefix + "Bypass");
+    var teamBandwidthElement = $(teamElementPrefix + "Bandwidth");
```
Comment (patfair) [inline static/js/field_monitor_display.js:38]: If there's room some indication of units might be good here too.

File: static/js/field_monitor_display.js [hunk 7]
```diff
@@ -118,6 +136,32 @@ var handleArenaStatus = function(data) {
...
+var handleMatchTime = function(data) {
+  translateMatchTime(data, function(matchState, matchStateText, countdownSec) {
+    $("#matchState").text(matchStateText);
+    $("#matchTime").text(countdownSec);
+    });
+};
+
+// Handles a websocket message to update the match score.
+var handleRealtimeScore = function(data,reversed) {
+
```
Comment (patfair) [inline static/js/field_monitor_display.js:149]: Remove unnecessary empty line

File: static/js/field_monitor_display.js [hunk 8]
```diff
@@ -118,6 +136,32 @@ var handleArenaStatus = function(data) {
...
+    if (reversed === "true") {
+      $("#rightScore").text(data.Red.ScoreSummary.Score);
+      $("#leftScore").text(data.Blue.ScoreSummary.Score);
+    } else {
+      $("#rightScore").text(data.Blue.ScoreSummary.Score);
+      $("#leftScore").text(data.Red.ScoreSummary.Score);
+
```
Comment (patfair) [inline static/js/field_monitor_display.js:156]: Remove unnecessary empty line

File: templates/field_monitor_display.html [hunk 9]
```diff
@@ -61,6 +69,7 @@
       </div>
       <div id="{{.side}}Team{{.position}}Robot" class="team-box center"
           title="Battery Voltage&#10;Seconds Since Last Connected"></div>
+      <!-- <div id="{{.side}}Team{{.position}}Bandwidth" class="team-box center" title="Team bandwidth"></div> -->
```
Comment (patfair) [inline templates/field_monitor_display.html:72]: Is this needed?

**Review by patfair — CHANGES_REQUESTED** [review]
Thanks; this looks useful.

File: static/js/field_monitor_display.js (hunk 4 above)
Comment (FletcherS7) [reply static/js/field_monitor_display.js:20]: Not always after Cheesy is stopped and started.

File: templates/field_monitor_display.html (hunk 9 above)
Comment (FletcherS7) [reply templates/field_monitor_display.html]: Uncommented

**Comment by FletcherS7** [comment]
@patfair I think I resolved all the changes you requested, also finalized the addition of the Bandwidth monitor

**Pull request merged**